File: 2015/day20/day20.py
import math

def giftsCalculatorPartOne(number):
    sum = 0
    sqrt_num = int(math.sqrt(number))
    for i in range(1, sqrt_num + 1):
        if number % i == 0:
            sum += i * 10
            if i != number // i:  
                sum += (number // i) * 10
    return sum

def giftsCalculatorPartTwo(number, elves):
    sum = 0
    sqrt_num = int(math.sqrt(number))
    for i in range(1, sqrt_num + 1):
        if number % i == 0:
            if elves[i] < 50:
                sum += i * 11
                elves[i] += 1
            if i != number // i and elves[number // i] < 50:
                sum += (number // i) * 11
                elves[number // i] += 1
    return sum

def findHousePartOne(target_gifts):
    houseNumber = 1
    while True:
        giftsPerHouse = giftsCalculatorPartOne(houseNumber)
        #print(f"Gifts for house {houseNumber} is {giftsPerHouse}")
        if giftsPerHouse >= target_gifts:
            break
        houseNumber += 1
    return houseNumber

def findHousePartTwo(target_gifts):
    houseNumber = 1
    elves = [0] * (target_gifts + 1) 
    while True:
        giftsPerHouse = giftsCalculatorPartTwo(houseNumber, elves)
        #print(f"Gifts for house {houseNumber} is {giftsPerHouse}")
        if giftsPerHouse >= target_gifts:
            break
        houseNumber += 1
    return houseNumber

number = 36000000
print(f"Part One: {findHousePartOne(number)}")
print(f"Part Two: {findHousePartTwo(number)}")

File: 2015/day20/day20.cpp
#include <iostream>
#include <cmath>
#include <vector>
using namespace std;

int giftsCalculatorPartOne(int number) {
    int sum = 0;
    int squareRoot = sqrt(number);
    for (int i = 1; i <= squareRoot; i++) {
        if (number % i == 0) {
            sum += i * 10;
            if (i != number / i) {
                sum += (number / i) * 10;
            }
        }
    }
    return sum;
}

int giftsCalculatorPartTwo(int number, vector<int>& elves) {
    int sum = 0;
    int squareRoot = sqrt(number);
    for (int i = 1; i <= squareRoot; i++) {
        if (number % i == 0) {
            if (elves[i] < 50) {
                sum += i * 11;
                elves[i]++;
            }
            int floorValue = number / i;
            if (i != floorValue && elves[floorValue] < 50) {
                sum += floorValue * 11;
                elves[floorValue]++;
            }
        }
    }
    return sum;
}

int findHousePartOne(int targetGifts) {
    int houseNumber = 1;
    while (true) {
        int giftsPerHouse = giftsCalculatorPartOne(houseNumber);
        if (giftsPerHouse >= targetGifts) {
            break;
        }
        houseNumber++;
    }
    return houseNumber;
}

int findHousePartTwo(int targetGifts) {
    int houseNumber = 1;
    vector<int> elves(targetGifts + 1, 0);  
    while (true) {
        int giftsPerHouse = giftsCalculatorPartTwo(houseNumber, elves);
        if (giftsPerHouse >= targetGifts) {
            break;
        }
        houseNumber++;
    }
    return houseNumber;
}

int main() {
    int number = 36000000;
    cout << "Part One: " << findHousePartOne(number) << endl;
    cout << "Part Two: " << findHousePartTwo(number) << endl;
    return 0;
}
